src/psyclone/parse/module_info.py

```diff
@@ -0,0 +1,212 @@
+    '''
+    PSyclone-specific exception for use when an error with the module manager
+    happens - typically indicating that some module information cannot be
+    found.
+
+    :param str value: the message associated with the error.
+
+    '''
+    def __init__(self, value):
+        PSycloneError.__init__(self, value)
+        self.value = "ModuleInfo error: "+str(value)
+
+
+# ============================================================================
+class ModuleInfo:
+    '''This class stores mostly cached information about modules: it stores
+    the original filename, if requested it will read the file and then caches
+    the plain text file, and if required it will parse the file, and then
+    cache the fparser AST.
+
+    :param str name: the module name.
+    :param str filename: the name of the source file that stores this module \
+        (including path).
+
+    '''
+
+    def __init__(self, name, filename):
+        self._name = name
+        self._filename = filename
+        # A cache for the source code:
+        self._source_code = None
+
+        # A cache for the fparser tree
+        self._parse_tree = None
+
+        # A cache for the module dependencies: this is just a set
+        # of all modules used by this module. Type: Set[str]
+        self._used_modules = None
+
+        # This is a dictionary, with the modules from the list of all used
+        # modules as key, and it stores the set of all symbols imported from
+        # this module: Dict[str, Set(str)]
+        self._used_symbols_from_module = None
+
+    # ------------------------------------------------------------------------
+    @property
+    def filename(self):
+        ''':returns: the filename that contains the source code for this \
+            module.
+        :rtype: str
+
+        '''
+        return self._filename
+
+    # ------------------------------------------------------------------------
+    def get_source_code(self):
+        '''Returns the source code for the module. The first time, it
+        will be read from the file, but the data is then cached.
+
+        :returns: the source code as string.
+        :rtype: str
+
+        :raises ModuleInfoError: when the file cannot be read.
+
+        '''
+        if self._source_code is None:
+            try:
+                with open(self._filename, "r", encoding='utf-8') as file_in:
+                    self._source_code = file_in.read()
+            except FileNotFoundError as err:
+                raise ModuleInfoError(
+                    f"Could not find file '{self._filename}' when trying to "
+                    f"read source code for module '{self._name}'") from err
+
+        return self._source_code
+
+    # ------------------------------------------------------------------------
+    def get_parse_tree(self):
+        '''Returns the fparser AST for this module. The first time, the file
+        will be parsed by fpaser, then the AST is cached for any future uses.
```

Comment:
Probably worth specifying that the Fortran2008 standard is used.

Reply:
OK. Is the return type still ok (:rtype: :py:class:`fparser.two.Fortran2003.Program`)? I believe so, but it just looks odd.

Reply:
Yes, it does look odd and yes, it's correct. Perhaps with some jiggery-pokery in fparser we could make e.g. `Program` available from `Fortran2008` as well as from `Fortran2003`. (I've just tried and currently `Fortran2008.Program` doesn't exist.)